src/Workspaces/Remote/Core/BrokeredServiceConnection.cs

```diff
@@ -293,22 +293,33 @@ public override async ValueTask<Optional<TResult>> TryInvokeAsync<TResult>(Solut
             //  6. 'pipe.Writer' is completed
             //  7. 'pipe.Reader' is completed
             //  8. OperationCanceledException is thrown back to the caller
+            //
+            // Since StreamJsonRpc will automatically complete writers when an error occurs (and cancellation is
+            // treated as an error), we create a buffered reader to isolate callback actions from this
+            // implementation.
 
             // Create a separate cancellation token for the reader, which we keep open until after the call to invoke
             // completes. If we close the reader before cancellation is processed by the remote call, it might block
             // (deadlock) while writing to a stream which is no longer processing data.
             using var readerCancellationSource = new CancellationTokenSource();
 
-            var pipe = new Pipe();
+            var writeBufferPipe = new Pipe();
+            var readBufferPipe = new Pipe();
 
             // Create new tasks that both start executing, rather than invoking the delegates directly
             // to make sure both invocation and reader start executing and transfering data.
 
             var writerTask = Task.Run(async () =>
             {
+                var copyTask = writeBufferPipe.Reader.CopyToAsync(readBufferPipe.Writer, cancellationToken);
+
                 try
                 {
-                    await invocation(service, pipe.Writer, cancellationToken).ConfigureAwait(false);
```

Comment:
@AArnott it's difficult to tell from the documentation, but it appears `pipe.Writer` will automatically be completed by StreamJsonRpc in the event of cancellation. This violates the ordering requirement above by allowing (6) to occur before (4).

Reply:
typically yes, but it's not a bad idea for you to complete these as well in your catch blocks. See my comment earlier in this file for more details.

Reply:
➡️ Now rewritten